chia/wallet/nft_wallet/nft_off_chain.py

```diff
@@ -0,0 +1,121 @@
+    # Read metadata from disk cache
+    file_name = get_cached_filename(nft_coin_info.nft_id, config)
+    if not file_name.exists():
+        return None
+    try:
+        text = file_name.read_text()
+        if verify_metadata(text, nft_coin_info.full_puzzle):
+            return text
+        else:
+            log.warning(f"Invalid cached metadata, NFT ID: {nft_coin_info.nft_id.hex()}")
+            return None
+    except Exception:
+        log.exception(f"Cannot load cached metadata of {nft_coin_info.nft_id.hex()}.")
+        return None
+
+
+def delete_off_chain_metadata(nft_id: bytes32, config: Optional[Dict[str, Any]] = None) -> None:
+    file_name = get_cached_filename(nft_id, config)
+    if file_name.exists():
+        file_name.unlink()
+        log.info(f"Deleted off-chain metadata of {nft_id.hex()}")
+
+
+def write_off_chain_metadata(nft_id: bytes32, metadata: str, config: Optional[Dict[str, Any]] = None) -> None:
+    file_name = get_cached_filename(nft_id, config)
+    file_name.write_text(metadata)
+
+
+async def get_off_chain_metadata(
+    nft_coin_info: NFTCoinInfo, config: Optional[Dict[str, Any]] = None, ignore_size_limit: bool = False
+) -> Optional[str]:
+    try:
+        # Check if the metadata is in disk cache
+        metadata = read_off_chain_metadata(nft_coin_info, config)
+        if metadata is not None:
+            return metadata
+        log.debug(f"{nft_coin_info.nft_id.hex()} is not in cache, downloading now ...")
+        metadata = await fetch_off_chain_metadata(nft_coin_info, ignore_size_limit)
+        if metadata is None:
+            log.error(f"Cannot find off-chain metadata of {nft_coin_info.nft_id.hex()}.")
+            return None
+        write_off_chain_metadata(nft_coin_info.nft_id, metadata, config)
+        log.info(f"Loaded off-chain metadata of {nft_coin_info.nft_id.hex()}")
+        return metadata
+    except Exception:
+        log.exception(f"Cannot get off-chain metadata of {nft_coin_info.nft_id.hex()}.")
+        return None
+
+
+def verify_metadata(metadata: str, full_puzzle: Program) -> bool:
+    uncurried_nft: Optional[UncurriedNFT] = UncurriedNFT.uncurry(*full_puzzle.uncurry())
+    if uncurried_nft is None:
+        return False
+    if uncurried_nft.meta_hash.as_python().hex() == std_hash(str.encode(metadata)).hex():
+        return True
+    else:
+        log.error(
+            f"Expect metadata hash: {uncurried_nft.meta_hash.as_python().hex()},"
+            f" actual metadata hash {std_hash(str.encode(metadata)).hex()}"
+        )
+        return False
+
+
+def get_cached_filename(nft_id: bytes32, config: Optional[Dict[str, Any]]) -> Path:
+    if config is None:
+        cache_path = DEFAULT_CACHE_PATH
+        hash_length = DEFAULT_PREFIX_HASH_LENGTH
+    else:
+        cache_path = str(config.get(CACHE_PATH_KEY, DEFAULT_CACHE_PATH))
+        hash_length = int(config.get(PREFIX_HASH_LENGTH_KEY, DEFAULT_PREFIX_HASH_LENGTH))
```

Comment:
Shouldn't these already be of these types anyways?

```suggestion
        cache_path: str = config.get(CACHE_PATH_KEY, DEFAULT_CACHE_PATH)
        hash_length: int = config.get(PREFIX_HASH_LENGTH_KEY, DEFAULT_PREFIX_HASH_LENGTH)
```